lib/api/api.js

```diff
@@ -42,47 +45,37 @@ const api = {
                 return callback(errors.InvalidArgument);
             }
             // Pull the source bucket and source object separated by /
-            const sourceBucket = source.slice(0, slashSeparator);
-            const sourceObject = source.slice(slashSeparator + 1);
-            const getRequestContext = new RequestContext(request.headers,
-                request.query, sourceBucket, sourceObject,
-                request.socket.remoteAddress, request.connection.encrypted,
-                'objectGet', 's3', locationConstraint);
-
-            return auth.doAuth(request, log, err => {
-                if (err) {
-                    log.trace('auth error on get portion of request');
-                    return callback(err);
-                }
-                const putRequestContext = new RequestContext(request.headers,
-                    request.query, request.bucketName, request.objectKey,
-                    request.socket.remoteAddress, request.connection.encrypted,
-                    'objectPut', 's3', locationConstraint);
-                return auth.doAuth(request, log, (err, authInfo) => {
-                    if (err) {
-                        log.trace('auth error on put portion of request');
-                        return callback(err);
-                    }
-                    return objectCopy(authInfo, request, sourceBucket,
-                        sourceObject, log, callback);
-                }, 's3', putRequestContext);
-            }, 's3', getRequestContext);
+            sourceBucket = source.slice(0, slashSeparator);
+            sourceObject = source.slice(slashSeparator + 1);
         }
-        const requestContext = new RequestContext(request.headers,
-            request.query, request.bucketName, request.objectKey,
-            request.socket.remoteAddress, request.connection.encrypted,
-            apiMethod, 's3', locationConstraint);
-        return auth.doAuth(request, log, (err, authInfo) => {
+        const requestContexts = prepareRequestContexts(apiMethod,
+            request, locationConstraint, sourceBucket, sourceObject);
+        return auth.doAuth(request, log, (err, userInfo,
+            authorizationResults) => {
             if (err) {
                 log.trace('authentication error', { error: err });
                 return callback(err);
             }
+            // TODO: When implement batch delete, check apiMethod
+            // and handle 'Deny' on key by key basis
+            if (authorizationResults) {
+                for (let i = 0; i < authorizationResults.length; i++) {
+                    if (!authorizationResults[i].isAllowed) {
+                        log.trace('authorization denial from Vault');
```

Comment:
It would be helpful to log which resource was denied authorization, do we have the info here?